discovery/marathon/marathon.go

```diff
@@ -435,7 +436,11 @@ func targetsForApp(app *App) []model.LabelSet {
 
 		for i := 0; i < len(app.PortDefinitions); i++ {
 			labels[i] = app.PortDefinitions[i].Labels
-			ports[i] = app.PortDefinitions[i].Port
+			// When requirePorts is false, this port becomes the 'servicePort', not the listen port.
+			// In this case, the port needs to be taken from the task instead of the app.
+			if app.RequirePorts {
```

Comment:
You should prefer relabelling to adding business logic to SD. If there's metadata missing you should add it.

Reply:
@brian-brazil This is the label/port gathering stage. The 'port' field means either 'data' port or 'service' port based on the state of this bool.

While I agree with your stance, I can't think of a way to handle this in relabeling without complicating the user's relabeling rule set. Additionally, the user will need to be aware of this idiosyncracy, one that we, the people working on this feature, were not even aware of until this issue popped up.

From the point of view of reducing maintenance cost and amount of incidents/issues, you are completely right, but this way we can at least fix it once, for all users. Unfortunately, consuming this API is hard to get right..

Reply:
The fact that you need to add a boolean configuration option that the user has to care about indicates that this belongs in relabelling. It's better to have everything in relabelling, than each SD slowly reinventing it.

Reply:
Hmm, `RequirePorts` is returned by the Marathon API, not exposed in the SD configuration if this is what you meant.

Reply:
@brian-brazil This is not a configuration option, it's a variable set in the Marathon app definition, obtained by querying the Marathon API.

Reply:
Ah, I'd missed that.

Reply:
No worries, thanks for taking the time to review. :+1: